NetKet/Machine/fullconn_layer.hpp

```diff
@@ -0,0 +1,257 @@
+
+  void GetParameters(VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(pars.data() + k, bias_.data(), out_size_ * scalar_bytesize_);
+      k += out_size_;
+    }
+
+    std::memcpy(pars.data() + k, weight_.data(),
+                in_size_ * out_size_ * scalar_bytesize_);
+  }
+
+  void SetParameters(const VectorType &pars, int start_idx) override {
+    int k = start_idx;
+
+    if (usebias_) {
+      std::memcpy(bias_.data(), pars.data() + k, out_size_ * scalar_bytesize_);
+
+      k += out_size_;
+    }
+
+    std::memcpy(weight_.data(), pars.data() + k,
+                in_size_ * out_size_ * scalar_bytesize_);
+  }
+
+  void InitLookup(const Eigen::VectorXd &v, LookupType &lt) override {
+    if (lt.VectorSize() == 0) {
+      lt.AddVector(out_size_);
+    }
+    lt.V(0) = (weight_.transpose() * v + bias_);
+  }
+
+  void UpdateLookup(const Eigen::VectorXd &v, const std::vector<int> &tochange,
+                    const std::vector<double> &newconf,
+                    LookupType &lt) override {
+    if (tochange.size() != 0) {
+      for (std::size_t s = 0; s < tochange.size(); s++) {
+        const int sf = tochange[s];
+        lt.V(0) += weight_.row(sf) * (newconf[s] - v(sf));
+      }
+    }
+  }
+
+  void Forward(const VectorType &prev_layer_data) override {
+    // Linear term z = W' * in + b
+    z_ = bias_;
+    z_.noalias() += weight_.transpose() * prev_layer_data;
+
+    // Apply activation function
+    activation_(z_, a_);
+  }
+
+  // Using lookup
+  void Forward(const VectorType & /*prev_layer_data*/,
+               const LookupType &lt) override {
+    z_ = lt.V(0);
+    // Apply activation function
+    activation_(z_, a_);
+  }
+
+  VectorType Output() const override { return a_; }
+
+  void Backprop(const VectorType &prev_layer_data,
+                const VectorType &next_layer_data, VectorType &der,
+                int start_idx) override {
+    // After forward stage, m_z contains z = W' * in + b
+    // Now we need to calculate d(L) / d(z) = [d(a) / d(z)] * [d(L) / d(a)]
+    // d(L) / d(a) is computed in the next layer, contained in next_layer_data
+    // The Jacobian matrix J = d(a) / d(z) is determined by the activation
+    // function
+    VectorType &dLz = z_;
+    activation_.ApplyJacobian(z_, a_, next_layer_data, dLz);
+
+    // Now dLz contains d(L) / d(z)
+    // Derivative for bias, d(L) / d(b) = d(L) / d(z)
+    int k = start_idx;
+
+    if (usebias_) {
+      Eigen::Map<VectorType> der_b{der.data() + k, out_size_};
+
+      der_b.noalias() = dLz;
+      k += out_size_;
+    }
+
+    // Derivative for weights, d(L) / d(W) = [d(L) / d(z)] * in'
+    Eigen::Map<MatrixType> der_w{der.data() + k, in_size_, out_size_};
+
+    der_w.noalias() = prev_layer_data * dLz.transpose();
+
+    // Compute d(L) / d_in = W * [d(L) / d(z)]
+    din_.noalias() = weight_ * dLz;
+  }
+
+  const VectorType &BackpropData() const override { return din_; }
```

Comment:
I still think that we should not have this BackpropData, and that din_ should just be returned by the Backprop method. Is there some case where this wouldn't work?